### ISSUE TITLE
caddy.Restart() - listen tcp :443: bind: address already in use

### ISSUE THREAD

**Issue by rushmorem** [mholt/caddy#661]
#### 1. What version of Caddy are you running (`caddy -version`)?
`9e386fc (latest commit on master at the moment)`

#### 2. What are you trying to do?
I'm using Caddy as a library to embed it so I can easily pull data from my database system, rebuild the Caddyfile and call `caddy.Restart(Caddyfile)` to restart gracefully. I love Caddy, but I have to say this ability to use Caddy as a library is by far my favourite feature :) It makes things that are a pain to do otherwise easy and straight forward.

#### 3. What is your entire Caddyfile?
My entire Caddyfile is huge. It has thousands of virtual hosts in it. In any case the file itself doesn't seem to have a problem because Caddy starts and runs fine when using it. The problem arises when a `caddy.Restart(Caddyfile)` is issued.

#### 4. How did you run Caddy (give the full command and describe the execution environment)?
I'm running Caddy as an embedded library. Here is my entire `main` function:-
```go
package main

import (
        "github.com/mholt/caddy/caddy"
        "log"
)

func init() {
        caddy.AppName = "name"
        caddy.AppVersion = "version"
}

func main() {
        // Start services
        cfg := new(Config)
        cfg.compile()
        err := caddy.Start(cfg)
        if err != nil {
                log.Fatal(err)
        }
        // Stop services on exit
        defer func() {
                err = caddy.Stop()
                if err != nil {
                        log.Fatal(err)
                }
        }()
        go cfg.watchUpdates()
        // Wait for services as they run
        caddy.Wait()
}
```

`watchUpdates()` waits for updates in the database and then compiles the Caddyfile and restarts the server:-
```go
c.compile()
caddy.Restart(c)
```
`c` is an instance of `Config`, an alias type of [caddyfile.Caddyfile](https://godoc.org/github.com/mholt/caddy/caddy/caddyfile#Caddyfile)

#### 5. What did you expect to see?
I expected the server to gracefully restart and the new Caddyfile to be applied after that.

#### 6. What did you see instead (give full error messages and/or log)?
I see either:-
`listen tcp :443: bind: address already in use`
or
`listen tcp :80: bind: address already in use`
in the logs. Also the old Caddyfile is still in use at this point which confirms that the restart is failing.

[caddy.Start()](https://godoc.org/github.com/mholt/caddy/caddy#Start) carries the following note:-
```go
// Note (POSIX): If Start is called in the child process of a
// restart more than once within the duration of the graceful
// cutoff (i.e. the child process called Start a first time,
// then called Stop, then Start again within the first 5 seconds
// or however long GracefulTimeout is) and the Caddyfiles have
// at least one listener address in common, the second Start
// may fail with "address already in use" as there's no
// guarantee that the parent process has relinquished the
// address before the grace period ends.
```

Inside the function, a couple of lines after that note I also see the following comment:-
```go
signalSuccessToParent() // TODO: Is doing this more than once per process a bad idea? Start could get called more than once in other apps.
```

In my logs after `[INFO] Restarting` I see messages printed to screen twice, not once so I suspect that the above is what's affecting me. That is `caddy.Start()` is somehow being called more than once during `caddy.Restart()`.

**Comment by mholt** [comment]
First, this is an _awesome_ issue. Thanks for using the new template!

One thing to be aware of when calling Restart() is that your Go process will spawn a new one, and the parent process will die, either immediately or in a few seconds.

That TODO next to signalSuccessToParent() might be outdated (oops), since the signalSuccessToParent() function uses a sync.Once to ensure that it only happens once, even if the function is called multiple times.

The reason your restart might be using the old Caddyfile is because you need to call [LoadCaddyfile()](https://godoc.org/github.com/mholt/caddy/caddy#LoadCaddyfile) in order for it to use the new one that you passed into Restart().

The caddy package actually uses [caddy.Input](https://godoc.org/github.com/mholt/caddy/caddy#Input) as the type to represent a Caddyfile. caddyfile.Caddyfile is for the upcoming API which will be able to transmit the Caddyfile as parsed tokens in JSON format.

Try using LoadCaddyfile() and let me see your updated code if you still have the problem.

**Comment by rushmorem** [comment]
I don't know how I overlooked `LoadCaddyfile()`. I was aware of `caddy.Input`. My `Config` was simply implementing it. I have since discovered `caddy.CaddyfileInput` so I'm now using that instead. My `Config` is still an alias type of `caddyfile.Caddyfile` though since I'm currently marshaling the struct to JSON and then using `caddyfile.FromJSON()` to convert it. I have been reading the wiki and studying the code. I can already see how I can improve my code. Thanks!
